include/kframework/lemmas/bytes-simplification.k

```diff
@@ -68,71 +68,158 @@ module BYTES-SIMPLIFICATION [symbolic]
-      requires 0 <=Int N1 andBool 0 <=Int N2
-      [simplification, concrete(N2), concrete(M)]
+    rule [range-reform-concat]:
+      #range(B1:Bytes, S:Int, minInt(W1:Int, W2:Int)) +Bytes ( #buf(maxInt(0, W3:Int), 0) +Bytes B2:Bytes ) => #range (B1, S, W1) +Bytes B2
+      requires W2 ==Int lengthBytes(B1) -Int S
+       andBool W3 ==Int S +Int W1 -Int lengthBytes(B1)
+       andBool 0 <=Int S andBool 0 <=Int W1 andBool 0 <=Int W2
+      [simplification]
 
-    rule #range(A +Bytes _B, START, LENGTH) => #range(A, START, LENGTH)
-      requires START +Int LENGTH <=Int lengthBytes(A) [simplification]
+    rule [range-buf-zero-1]:
+      #range(#buf(32, X:Int), 0, W:Int) => #buf(W, 0)
+      requires 0 <=Int W andBool W <=Int 32
+       andBool #rangeUInt (8 *Int (32 -Int W), X)
+      [simplification, concrete(W)]
 
-    rule #range(A +Bytes B, START, LENGTH) => #range(B, START -Int lengthBytes(A), LENGTH)
-      requires lengthBytes(A) <=Int START [simplification]
+    rule [range-buf-zero-2-symb]:
+      #range(#buf(W:Int, 0), S1:Int, W1:Int) => #buf(W1, 0)
+      requires 0 <=Int W andBool 0 <=Int S1 andBool 0 <=Int W1
+      [simplification]
 
-    rule #range(BA:Bytes, 0, LEN) => BA requires lengthBytes(BA) ==Int LEN [simplification]
+    rule [range-buf-zero-2-conc]:
+      #range(B, S1:Int, W1:Int) => #buf(W1, 0)
+      requires 0 <=Int S1 andBool 0 <=Int W1
+       andBool B ==K #buf(lengthBytes(B), 0)
+      [concrete(B), simplification]
+
+    rule [range-memUpdate-before]:
+      #range(B1:Bytes [ S1:Int := B2:Bytes ], S2:Int, W2:Int) =>
+        #let W = minInt(W2, S1 -Int S2) #in
+          #range(B1, S2, W) +Bytes #range(B1 [ S1 := B2 ], S1, W2 -Int W)
+      requires 0 <=Int S2 andBool S2 <Int S1
+      [simplification]
 
-    rule (#range(#buf(32, X), 0, 28) ):Bytes => #buf(28, 0) requires #rangeUInt ( 32 , X ) [simplification]
+    rule [range-memUpdate-in-between]:
+      #range(B1:Bytes [ S1:Int := B2:Bytes ], S2:Int, W2:Int) =>
+        #let S = S2 -Int S1 #in
+        #let W = minInt(W2, lengthBytes(B2) -Int S) #in
+          #range(B2, S, W) +Bytes #range(B1, S1 +Int lengthBytes(B2), W2 -Int W)
+      requires 0 <=Int S1 andBool S1 <=Int S2 andBool S2 <=Int S1 +Int lengthBytes(B2)
+      [simplification]
+
+    rule [range-memUpdate-outside]:
+      #range(B1:Bytes [ S1:Int := B2:Bytes ], S2:Int, W2:Int) => #range(B1, S2, W2)
+      requires 0 <=Int S1 andBool 0 <=Int W2 andBool S1 +Int lengthBytes(B2) <=Int S2 [simplification]
 
-    rule (#range(#range(BA:Bytes, S0, L0), S1, L1)) => #range(BA, S0 +Int S1, minInt(L1, L0 -Int S1)) +Bytes #buf(maxInt(0, L1 -Int (L0 -Int S1)), 0)
-      requires 0 <=Int S0 andBool 0 <=Int L0 andBool 0 <=Int S1 andBool 0 <=Int L1 [simplification]
+    rule [range-buf-zero]:
+      #range (#buf(W1:Int, X:Int), S2:Int, W2:Int) => #buf(W2, 0)
+      requires 0 <=Int X andBool X <Int 2 ^Int (8 *Int (W1 -Int (S2 +Int W2))) andBool S2 +Int W2 <=Int W1
+      [simplification]
```

Comment:
Counterexample:

- `W1 == 8`
- `X == 1`
- `S2 == -3`
- `W2 == 1`

Then:

- `0 <=Int X => true`
- `X <Int 2 ^Int (8 *Int (W1 -Int (S2 +Int W2))) => 1 <Int 2 ^Int (8 *Int (8 -Int (-3 +Int 1))) => 1 <Int 2 ^Int (8 *Int 10) => true`
- `S2 +Int W2 <=Int W1 => -3 +Int 1 <=Int 8 => true`
- `#range(#buf(W1, X), S2, W2) => #range(#buf(W1, X), -3, 1) => .Bytes`
- `#range(#buf(W1, X), S2, W2) => #range(#buf(8, 1), -3, 1) => #buf(1, 0)`

I think conditions `0 <=Int S2 andBool 0 <=Int W2` should be good?

Reply:
I think `0 <=Int S2` is sufficient because if `W2 <Int 0` then both sides collapse to `.Bytes`, but I'm happy to add both for sanity.